File: lib/languages/zig/loops.zig
// For loop with range
fn countUp(n: i32) void {
    for (1..n + 1) |i| {
        std.debug.print("{d}\n", .{i});
    }
}

// While loop
fn countDown(from: i32) void {
    var current = from;
    while (current > 0) {
        std.debug.print("{d}\n", .{current});
        current -= 1;
    }
}

// Infinite loop with break
fn findFirstEven(numbers: []const i32) ?i32 {
    var index: usize = 0;
    while (true) {
        if (index >= numbers.len) {
            break null;
        }
        
        if (numbers[index] % 2 == 0) {
            break numbers[index];
        }
        
        index += 1;
    }
}

// For loop with array
fn sumArray(numbers: []const i32) i32 {
    var sum: i32 = 0;
    for (numbers) |num| {
        sum += num;
    }
    return sum;
}

// For loop with index
fn printIndexed(items: []const []const u8) void {
    for (items, 0..) |item, index| {
        std.debug.print("{d}: {s}\n", .{index, item});
    }
}

// Break and continue
fn sumUntilNegative(numbers: []const i32) i32 {
    var sum: i32 = 0;
    for (numbers) |num| {
        if (num < 0) {
            break;
        }
        if (num == 0) {
            continue;
        }
        sum += num;
    }
    return sum;
}

// Labeled breaks
fn findInMatrix(matrix: []const []const i32, target: i32) ?struct { usize, usize } {
    outer: for (matrix, 0..) |row, i| {
        for (row, 0..) |value, j| {
            if (value == target) {
                return .{ i, j };
            }
        }
    }
    return null;
}

// While with optional unwrapping
fn processQueue(queue: *std.ArrayList(i32)) void {
    while (queue.popOrNull()) |value| {
        std.debug.print("Processing: {d}\n", .{value});
    }
}

// Comptime for loops (executed at compile time)
fn comptimeSum(comptime numbers: []const i32) i32 {
    comptime {
        var sum: i32 = 0;
        for (numbers) |num| {
            sum += num;
        }
        return sum;
    }
}

// Inline for (unrolls the loop)
fn inlineSum(numbers: []const i32) i32 {
    var sum: i32 = 0;
    inline for (numbers) |num| {
        sum += num;
    }
    return sum;
}

File: lib/languages/rust/loops.rs
// For loop with range
fn count_up(n: i32) {
   for i in 1..=n {
       println!("{}", i);
   }
}

// While loop
fn count_down(from: i32) {
   let mut current = from;
   while current > 0 {
       println!("{}", current);
       current -= 1;
   }
}

// Loop (infinite) with break
fn find_first_even(numbers: &[i32]) -> Option<i32> {
   let mut index = 0;
   loop {
       if index >= numbers.len() {
           break None;
       }
       
       if numbers[index] % 2 == 0 {
           break Some(numbers[index]);
       }
       
       index += 1;
   }
}

// For loop with iterator
fn sum_array(numbers: &[i32]) -> i32 {
   let mut sum = 0;
   for num in numbers {
       sum += num;
   }
   sum
}

// For loop with enumerate
fn print_indexed(items: &[String]) {
   for (index, item) in items.iter().enumerate() {
       println!("{}: {}", index, item);
   }
}

// Break and continue
fn sum_until_negative(numbers: &[i32]) -> i32 {
   let mut sum = 0;
   for &num in numbers {
       if num < 0 {
           break;
       }
       if num == 0 {
           continue;
       }
       sum += num;
   }
   sum
}

// Labeled breaks
fn find_in_matrix(matrix: &[&[i32]], target: i32) -> Option<(usize, usize)> {
   'outer: for (i, row) in matrix.iter().enumerate() {
       for (j, &value) in row.iter().enumerate() {
           if value == target {
               return Some((i, j));
           }
       }
   }
   None
}

// While let loop (pattern matching)
fn process_queue(mut queue: Vec<i32>) {
   while let Some(value) = queue.pop() {
       println!("Processing: {}", value);
   }
}

// Functional approaches
fn double_all(numbers: &[i32]) -> Vec<i32> {
   numbers.iter().map(|&x| x * 2).collect()
}

fn only_positive(numbers: &[i32]) -> Vec<i32> {
   numbers.iter().filter(|&&x| x > 0).copied().collect()
}

